homeassistant/components/systemmonitor/sensor.py

```diff
@@ -158,183 +181,289 @@ def check_required_arg(value):
+        entities.append(SystemMonitorSensor(sensor_registry, type_, argument))
+
+    scan_interval = config.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    await async_setup_sensor_registry_updates(hass, sensor_registry, scan_interval)
+
+    async_add_entities(entities)
+
+
+async def async_setup_sensor_registry_updates(
+    hass: HomeAssistant,
+    sensor_registry: dict[str, SensorData],
+    scan_interval: datetime.timedelta,
+) -> None:
+    """Update the registry and create polling."""
+
+    _update_lock = asyncio.Lock()
+
+    def _update_sensors() -> None:
+        """Update sensors and store the result in the registry."""
+        for type_, data in sensor_registry.items():
+            try:
+                state, value, update_time = _update(type_, data)
+            except Exception as ex:  # pylint: disable=broad-except
+                _LOGGER.exception("Error updating sensor: %s", type_, exc_info=ex)
+                data.last_exception = ex
+            else:
+                data.state = state
+                data.value = value
+                data.update_time = update_time
+                data.last_exception = None
+
+    async def _async_update_data(*_: Any) -> None:
+        """Update all sensors in one executor jump."""
+        if _update_lock.locked():
+            _LOGGER.warning(
+                "Updating systemmonitor took longer than the scheduled update interval %s",
+                scan_interval,
+            )
+            return
+
+        async with _update_lock:
+            await hass.async_add_executor_job(_update_sensors)
+            async_dispatcher_send(hass, SIGNAL_SYSTEMMONITOR_UPDATE)
+
+    polling_remover = async_track_time_interval(hass, _async_update_data, scan_interval)
+
+    @callback
+    def _async_stop_polling(*_: Any) -> None:
+        polling_remover()
 
-    add_entities(dev, True)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop_polling)
+
+    await _async_update_data()
 
 
 class SystemMonitorSensor(SensorEntity):
     """Implementation of a system monitor sensor."""
 
-    def __init__(self, sensor_type, argument=""):
+    def __init__(
+        self,
+        sensor_registry: dict[str, SensorData],
+        sensor_type: str,
+        argument: str = "",
+    ) -> None:
         """Initialize the sensor."""
-        self._name = "{} {}".format(SENSOR_TYPES[sensor_type][0], argument)
-        self._unique_id = slugify(f"{sensor_type}_{argument}")
-        self.argument = argument
-        self.type = sensor_type
-        self._state = None
-        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
-        self._available = True
-        if sensor_type in ["throughput_network_out", "throughput_network_in"]:
-            self._last_value = None
-            self._last_update_time = None
+        self._type: str = sensor_type
+        self._name: str = "{} {}".format(self.sensor_type[SENSOR_TYPE_NAME], argument)
```

Comment:
We can use an f-string.

Reply:
I didn't clean up any of the existing code beyond adding typing.  I'll do that later today